### PR TITLE
[Lotto - Step1] Submit bojana-petroska's mission

### PR DESCRIPTION
**Pull request by bojana-petroska**
#### Lotto Step 1
This PR implements several features for the **Lotto Game**, including: `Input View`, `ResultView`, `LottoGame`.

#### Feedback Requested
- Is the separation of concerns properly applied, or would it be better to compress some logic to fewer files?
- Are the tests comprehensive enough, or should they be more detailed and edge case tests?
- Regarding error handling: would it make sense to add try & catch blocks inside the input handling?
- Could you please review the overall architecture?
- Suggestions for improving Kotlin best practices are also welcome.

Thank you very much for your time!
Bojana

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,34 @@
...
+- [x] machine issues lotto tickets
+- [x] figure out winning tickets
+
+#### Input
+- [x] get input of amount
+- [x] get the winning numbers
+- [x] get the bonus number
+
+#### Output result
+- [x] should print the number of purchased tickets
```
Comment (Hyeon9mak) [inline README.md:31]: Nice work! 👍

Organizing business or feature requirements in documents like the README.md(or something), will be helpful in the future whether it'syou revisiting the project later on, or a fellow developer joining the team and looking at the project for the first time.

Hope you keep this habit going and keep refining it!

File: src/main/kotlin/lotto/Application.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    val winningNumbers = Lotto(InputView.getWinningNumbers())
+    val bonusNumber = InputView.getBonusNumber()
+
+    val game = LottoGame(lottoTickets, winningNumbers, bonusNumber)
+
+    game.play()
+
+    ResultView.viewWinning(game.results)
+    ResultView.viewRate(game.winningAmount, purchaseAmount.amount)
+}
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/Application.kt]: `main` function is solely responsible for creating objects, controlling the flow, and communicating between Domain and View classes. Very clean and impressive! 👍 💯

File: src/main/kotlin/lotto/Lotto.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package lotto
+
+private const val NUMBERS_SIZE = 6
+
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/Lotto.kt]: what's the different between outer class const value and compaion object const value?

File: src/main/kotlin/lotto/Lotto.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+open class Lotto(val numbers: List<Int>) {
+    init {
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/Lotto.kt]: what reason of choose the `open class`? 🤔

File: src/main/kotlin/lotto/Lotto.kt [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        require(numbers.size == NUMBERS_SIZE) {
+            "Size of numbers should be equal $NUMBERS_SIZE"
+        }
+        require(numbers.toSet().size == NUMBERS_SIZE) {
+            "lotto ticket numbers should be unique"
+        }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/Lotto.kt]: Since duplicates aren’t allowed, how about using data structure better suited than a List? 😉

File: src/main/kotlin/lotto/LottoGame.kt [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package lotto
+
+class LottoGame(val playersTickets: List<Lotto>, val winningNumbers: Lotto, val bonusNumber: Int) {
+    val results: MutableMap<Rank, Int> = mutableMapOf()
+    var winningAmount = 0
+
+    fun play() {
+        winningAmount == 0
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/LottoGame.kt]: maybe... `winningAmount = 0` ?

Comment (Hyeon9mak) [reply src/main/kotlin/lotto/LottoGame.kt]: It looks like a new instance will be allocated every time a new game starts.
Since `winningAmount` and `results` will also be freshly initialized each time, there doesn't seem to be a need to initialize them in every `play()` call!

File: src/main/kotlin/lotto/LottoMachine.kt [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+    fun issueTickets(): List<Lotto> {
+        val ticketCount: Int = money.amount / TICKET_COST
+
+        return (1..ticketCount).map {
+            Lotto(generateRandomNumber())
+        }
+    }
+
+    private fun generateRandomNumber() = (1..45).shuffled().take(6).sorted()
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/LottoMachine.kt:13]: nice work! It looks very Kotlin-like 😄 👍

File: src/main/kotlin/lotto/LottoMachine.kt [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+class LottoMachine(val money: Money) {
+    val tickets: List<Lotto> = issueTickets()
+
+    fun issueTickets(): List<Lotto> {
+        val ticketCount: Int = money.amount / TICKET_COST
+
+        return (1..ticketCount).map {
+            Lotto(generateRandomNumber())
+        }
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/LottoMachine.kt]: how about this? 😉 
base on constructor(with `init`)

```suggestion
class LottoMachine(money: Money) {
    val tickets: List<Lotto>

    init {
        val ticketCount: Int = money.amount / TICKET_COST
        tickets = List(ticketCount) { Lotto(generateRandomNumber()) }
    }
```

File: src/main/kotlin/lotto/Money.kt [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+class Money(val amount: Int) {
+    init {
+        require(amount >= 1000) {
+            "the amount should be minumum 1,000 KRW"
+        }
+        require(amount % 1000 == 0) {
+            "the amount should be divisible by 1000"
+        }
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/Money.kt]: What are the differences between `value class`, `data class`, and `class`?

File: src/main/kotlin/lotto/Rank.kt [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    val matchBonus: Boolean,
+    val prizeAmount: Int,
+) {
+    FIRST(6, false, 2_000_000_000),
+    SECOND(5, true, 30_000_000),
+    THIRD(5, false, 1_500_000),
+    FOURTH(4, false, 50_000),
+    FIFTH(3, false, 5_000),
+    NONE(0, false, 0), ;
+
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/Rank.kt:14]: if `prizeAmount` changes or new rank are added, only need to look at the `Rank` enum class! 💯 👍

File: src/main/kotlin/view/InputView.kt [hunk 11]
```diff
@@ -0,0 +1,30 @@
+package view
+
+object InputView {
```
Comment (Hyeon9mak) [inline src/main/kotlin/view/InputView.kt:3]: it doesn't have any properties. it would be good to use the object class 👍 💯

File: src/main/kotlin/view/InputView.kt [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    fun getWinningNumbers(): List<Int> {
+        println("Please enter last week's winning numbers.")
+        val input = readLine() ?: ""
+        require(input.isNotBlank()) {
+            "Input cannot be empty."
+        }
+        return input.split(",").map { it.toIntOrNull() ?: 0 }
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/view/InputView.kt:20]: nice!
I think determining the criteria for splitting and even validation for blank should be the view's responsibility. Because when it changes from terminal text-based to web browser image representation, all of these would be affected.

File: src/test/kotlin/lotto/CalculatorTest.kt [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+class CalculatorTest {
+    @Test
+    fun `should divide integers and return a double`() {
+        val a = 5000
+        val b = 1000
+        val result = 5.00
+        assertThat(Calculator.division(a, b)).isEqualTo(result)
+    }
+}
```
Comment (Hyeon9mak) [inline src/test/kotlin/lotto/CalculatorTest.kt:14]: I was wondering 'why Bojana created a division object function?',
it was to make testing easier! That's brilliant! 😄

File: src/test/kotlin/lotto/LottoMachineTest.kt [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+
+private const val TICKET_COST = 1000
+
+class LottoMachineTest {
+    @Test
+    fun `machine should issue 2 tickets`() {
+        val lottoMachine = LottoMachine(Money(2000))
+        assertTrue(lottoMachine.tickets.size == 2000 / TICKET_COST)
+    }
+}
```
Comment (Hyeon9mak) [inline src/test/kotlin/lotto/LottoMachineTest.kt]: How about also testing with boundary values? like `2_001`, `1_999` or something.

File: src/test/kotlin/lotto/LottoTest.kt [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+        val lotto = Lotto(listOf(1, 2, 3, 4, 5, 6))
+        assertTrue(lotto.getListOfNumbers().size == NUMBERS_SIZE)
+    }
+
+    @Test
+    fun `if lotto ticket numbers are not unique thrown`() {
+        assertThrows<IllegalArgumentException> {
+            Lotto(listOf(1, 2, 3, 4, 5, 5))
+        }
+    }
```
Comment (Hyeon9mak) [inline src/test/kotlin/lotto/LottoTest.kt]: 👍 💯

File: src/test/kotlin/lotto/LottoTest.kt [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Test
+    fun `if lotto ticket each number is not in range between 1 and 45 thrown`() {
+        assertThrows<IllegalArgumentException> {
+            Lotto(listOf(56, 1, 2, 3, 4, 5))
+        }
+    }
```
Comment (Hyeon9mak) [inline src/test/kotlin/lotto/LottoTest.kt]: the functional requirement is `1 and 45`, so i think this is better test case.

```kotlin
1, 2, 3, 4, 5, 46
0, 2, 3, 4, 5, 6
```

File: src/main/kotlin/lotto/Application.kt [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+    val lottoMachine = LottoMachine(purchaseAmount)
+    val lottoTickets = lottoMachine.tickets
+
+    ResultView.viewTicketsAmount(lottoTickets.size)
+    ResultView.viewTickets(lottoTickets)
+
+    val winningNumbers = Lotto(InputView.getWinningNumbers())
+    val bonusNumber = InputView.getBonusNumber()
+
+    val game = LottoGame(lottoTickets, winningNumbers, bonusNumber)
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/Application.kt]: I think the `Lotto`'s six numbers and bonusNumber are the same thing.
so bonusNumber need validation too.

How about wrapping both of them with the same VO (Value Object) and moving the validation there?
like the `Money` class!

**Review by Hyeon9mak — CHANGES_REQUESTED** [review]
Hi Bojana!
It was readable and clean code 👍 
Below are the answers to your questions.

### Is the separation of concerns properly applied, or would it be better to compress some logic to fewer files?

What makes for good separation of concerns?
What criteria should we use to decide whether to split responsibilities further? 
I can provide an answer once Bojana establishes those criteria.

I believe the criteria for design should be business-driven, not technology-driven.
The design should be flexible enough to adapt to the business problems and characteristics that Bojana is trying to solve.

- What kind of design would be good for a business where Views change frequently but Domain changes are rare?
- What kind of design would be good for a business where Views don't change but Domain changes are frequent?

During the HeroTechCourse mission, how about setting the standard as "a business where Views change frequently, but Domain changes are rare"?
Or please establish your own criteria and explain the background to the reviewer!
I believe this will make for more enjoyable mission progress and code reviews.

### Are the tests comprehensive enough, or should they be more detailed and edge case tests?

I left detail comments on test codes 😉

### Regarding error handling: would it make sense to add try & catch blocks inside the input handling?

It's definitely good to add validation since all users behave unpredictably in unexpected ways. 
However, if u go too far with it, u might loss time to spend with your friends after work. 😄 
How about adding some simple validation that finds a reasonable balance point?

### Could you please review the overall architecture?

I think u already know(and good at) MVC architecture!
if u want more review for the overall architecture, plz re-question to next request. 😄 

Feel free to ask any questions you have!

File: src/main/kotlin/lotto/Lotto.kt (hunk 3 above)
Comment (bojana-petroska) [reply src/main/kotlin/lotto/Lotto.kt]: the outer const class value or top level const is a global variable, and the const from the inside of a companion object is tied to the class itself / scoped to the class itself. I see your point! Thanks a lot for the tip ; )

File: src/main/kotlin/lotto/Lotto.kt (hunk 4 above)
Comment (bojana-petroska) [reply src/main/kotlin/lotto/Lotto.kt]: we planned on doing an extension on the class but didn't do it and didn't remove the open keyword

File: src/main/kotlin/lotto/Lotto.kt (hunk 5 above)
Comment (bojana-petroska) [reply src/main/kotlin/lotto/Lotto.kt]: good point! I did some refactoring here : )

File: src/main/kotlin/lotto/Money.kt (hunk 9 above)
Comment (bojana-petroska) [reply src/main/kotlin/lotto/Money.kt]: Thanks for the tip! I used the value class because it is a wrapper for a single value with type safety. A data class would be more useful for more complex data where we can use it's built in functions. A Class is the most general type that holds state and behavior but doesn't provide built in functions. I refactored the Money class to a value class.

File: src/main/kotlin/view/InputView.kt (hunk 12 above)
Comment (bojana-petroska) [reply src/main/kotlin/view/InputView.kt:20]: do you suggest to separate the concerns or not particulary for this project because it is small in size?

File: src/test/kotlin/lotto/CalculatorTest.kt (hunk 13 above)
Comment (bojana-petroska) [reply src/test/kotlin/lotto/CalculatorTest.kt:14]: Thanks! Testing is easier and we separated the business logic (mathematical operations) from view.

File: src/test/kotlin/lotto/LottoMachineTest.kt (hunk 14 above)
Comment (bojana-petroska) [reply src/test/kotlin/lotto/LottoMachineTest.kt]: The Money value Class require for the amount to be divisible by 1000, so these inputs would be invalid. Or, you are suggesting me to make them as edge cases, and purposely return an error?

File: src/main/kotlin/lotto/Lotto.kt (hunk 5 above)
Comment (bojana-petroska) [reply src/main/kotlin/lotto/Lotto.kt]: after the rest of your review, I did another refactoring : ))

File: src/test/kotlin/lotto/BonusNumberTest.kt [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+import kotlin.test.assertEquals
+
+class BonusNumberTest {
+    @Test
+    fun `should create BonusNumber when number in range and not in lotto numbers`() {
+        val primaryNumbers = setOf(1, 2, 3, 4, 5, 6)
+        val bonus = BonusNumber.of(7, primaryNumbers)
+        assertEquals(7, bonus.number)
```
Comment (bojana-petroska) [inline src/test/kotlin/lotto/BonusNumberTest.kt:12]: this is wrong, I corrected it:
`assertThat(bonus.number).isEqualTo(7)`

File: src/main/kotlin/lotto/LottoNumbers.kt [hunk 19]
```diff
@@ -0,0 +1,21 @@
+package lotto
+
+@JvmInline
+value class LottoNumbers(val numbers: Set<Int>) {
+    companion object {
+        private const val NUMBERS_SIZE = 6
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/LottoNumbers.kt:6]: `LottoNumbers` VO 💯 👍 
now we can 100% trust value after `LottoNumbers` created.
and how about create more VO of `LottoNumber`?
it can make 100% trust value of the lotto number & bonus number

File: src/main/kotlin/lotto/LottoNumbers.kt [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+                "Size of numbers should be equal $NUMBERS_SIZE"
+            }
+            require(numbers.toSet().size == NUMBERS_SIZE) {
+                "lotto ticket numbers should be unique"
+            }
+            require(numbers.all { it in 1..45 }) {
+                "lotto ticket each number should be between 1 and 45"
+            }
+            return LottoNumbers(numbers.toSet())
+        }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/LottoNumbers.kt:19]: nice validations. it makes easy to generate test code. 👍 
how about move this validations to `LottoNumbers` init(constructor) area?
then we can get stability about `LottoNumbers` when it created by Set values also!

File: src/main/kotlin/view/InputView.kt (hunk 12 above)
Comment (Hyeon9mak) [reply src/main/kotlin/view/InputView.kt:20]: Sorry, I misunderstood your question.
My comment was meant to say that "the view-related logic is well organized in the InputView".
Do you have any other questions?

File: src/test/kotlin/lotto/LottoMachineTest.kt [hunk 21]
```diff
@@ -0,0 +1,14 @@
+package lotto
+
+import org.junit.jupiter.api.Assertions.assertTrue
+import org.junit.jupiter.api.Test
+
+private const val TICKET_COST = 1000
+
```
Comment (Hyeon9mak) [inline src/test/kotlin/lotto/LottoMachineTest.kt:7]: can be move to compaion object! 👀

File: src/test/kotlin/lotto/LottoMachineTest.kt [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+class LottoMachineTest {
+    @Test
+    fun `machine should issue 2 tickets`() {
+        val lottoMachine = LottoMachine(Money.of(2000))
+        assertTrue(lottoMachine.tickets.size == 2000 / TICKET_COST)
+    }
+}
```
Comment (Hyeon9mak) [inline src/test/kotlin/lotto/LottoMachineTest.kt:14]: > The Money value Class require for the amount to be divisible by 1000, so these inputs would be invalid. Or, you are suggesting me to make them as edge cases, and purposely return an error?

I think... the most of fellow developers will be wonder of LottoMachine's edge case.
And HeroTechCourse missions are designed to help you build good habits.
So how about append test little by little? 😉

File: src/main/kotlin/lotto/Rank.kt [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+            return when {
+                matchCount == FIRST.matchCount -> FIRST
+                matchCount == SECOND.matchCount && (SECOND.matchBonus == matchBonus) -> SECOND
+                matchCount == THIRD.matchCount -> THIRD
+                matchCount == FOURTH.matchCount -> FOURTH
+                matchCount == FIFTH.matchCount -> FIFTH
+                else -> NONE
+            }
+        }
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/Rank.kt:29]: It would be great to have tests for this, too!

**Review by Hyeon9mak — APPROVED** [review]
Hi Bojana, Great job the reflect the feedback!

As u know, I've added some super easy-comments. 😄 
You can go directly to the next step. See you in next step!

If you have more questions, plz feel free to leave a comment. 👍
